cachi2/core/package_managers/general.py

```diff
@@ -33,6 +44,172 @@ def download_binary_file(url, download_path, auth=None, insecure=False, chunk_si
+                f"Could not download {tarball_name} from {url}. "
+                f"exception_name: {exception.__class__.__name__}, "
+                f"details: {exception}"
+            )
+        ) from None
+
+    log.debug(f"Download completed - {tarball_name}")
+
+
+async def get_dependencies(
+    download_dir: Path,
+    deps_to_download: List[str],
+    concurrency_limit: int,
+) -> List[str]:
+    """
+    Asynchronous function that execute the dependencies download.
+
+    Receives the url (proxy_repo_url), the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param Path download_dir: Path to download file to.
+    :param list[str] deps_to_download: List of dependencies to be downloaded.
+    :param int concurrency_limit: Max number of concurrent tasks (downloads).
+    :return: a list of the downloaded tarballs.
+    :rtype: list[str]
+    """
+
+    attempts = 3
+
+    async def on_request_start(
+        session,
+        trace_config_ctx,
+        params,
+    ) -> None:
+        current_attempt = trace_config_ctx.trace_request_ctx["current_attempt"]
+        if current_attempt > 1:
+            package_name = str(params.url).split("/")[-1]
+            log.debug(f"Attempt {current_attempt}/{retry_options.attempts} - {package_name}")
+
+    trace_config = aiohttp.TraceConfig()
+    trace_config.on_request_start.append(on_request_start)
+    retry_options = aiohttp_retry.JitterRetry(attempts=attempts, retry_all_server_errors=True)
+    retry_client = aiohttp_retry.RetryClient(retry_options=retry_options, trace_configs=[trace_config])
+
+    async with retry_client as session:
+
+        tasks: Set[asyncio.Task] = set()
+
+        results = []
+
+        for dependency_url in deps_to_download:
+            log.info("Downloading - %s ", dependency_url)
+
+            if len(tasks) >= concurrency_limit:
+                # Wait for some download to finish before adding a new one
+                done, tasks = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+                # Check for exceptions
+                try:
+                    await asyncio.gather(*done)
+                except FetchError:
+                    # Close retry_client if any request fails (other tasks can be running,
+                    # if a task is closed with the client open, an Warning is raised).
+                    await retry_client.close()
+                    for t in tasks:
+                        t.cancel()
+                    raise
+
+            results.append(dependency_url)
+
+            if dependency_url[-4:] == ".tgz":
+                # Download tar archive
+                tasks.add(
+                    asyncio.create_task(
+                        async_download_binary_file(
+                            session,
+                            dependency_url,
+                            download_dir,
+                            dependency_url.split("/")[-1],
```

Comment:
Taking the part of the url after the last `/` as the filename is not quite correct

* resolved urls can have `#<hash>` fragments
* for scoped packages (`@{scope}/{name}`), the filename should be `{scope}-{name}-{version}.tgz`: https://github.com/containerbuildsystem/cachito/blob/db612e6022311644913c611e3553097490cc94b1/cachito/workers/pkg_managers/general_js.py#L77

It might be best to construct the download path from the package name and version, not from the url

We should also make sure that the filenames for git dependencies and https dependencies will never conflict with other dependencies, probably by using the same download paths as we do in Cachito (and the same as we use for pip) https://github.com/containerbuildsystem/cachi2/blob/a1944191430efe2fe158f3518017abe58c2a665a/cachi2/core/package_managers/pip.py#L1831

Reply:
is this desired way? https://github.com/containerbuildsystem/cachi2/pull/152/files#diff-1bc123673d60c878542386b0a8ba8a27e381c4627075c4a5f3e106f0e74263acR232

Reply:
I added special paths for git sources, PTAL. https://github.com/containerbuildsystem/cachi2/pull/152/files#diff-1bc123673d60c878542386b0a8ba8a27e381c4627075c4a5f3e106f0e74263acR202

```
/tmp/pytest-of-lkolacek/pytest-current/test_npm_smoketest_npm_smoketecurrent/npm_smoketest_simple-output/
└── deps
    └── npm
        ├── debug-4.3.4.tgz
        ├── git@github.com
        │   ├── cachito-testing
        │   │   └── cachito-npm-without-deps
        │   │       └── cachito-npm-without-deps-external-gitcommit-2f0ce1d7b1f8b35572d919428b965285a69583f6.tgz
        │   └── kevva
        │       └── is-positive
        │           └── is-positive-external-gitcommit-97edff6f525f192a3f83cea1944765f769ae2678.tgz
        └── ms-2.1.2.tgz

```

Reply:
`git@github.com` should be `github.com` (maybe use `parsed_url.hostname` rather than the netloc)